### PR TITLE
Stretched WENO and general improvement of WENO infrastructure

### LINKED ISSUE #1704: Weighted WENO stencils for stretched grids
In [A simple algorithm to improve the performance of the WENO scheme on non-uniform grids](https://link.springer.com/article/10.1007/s10409-017-0715-2) a simple method for generalizing WENO stencils for stretched grids is described. This scheme is easy to implement and limits to what we already have implemented in the case of a regular grid.

### PR DESCRIPTION
**Pull request by simone-silvestri**
Extended the type WENO5 to allow the execution of the WENO scheme on stretched and uniform grids alike. 

In practice:

- `WENO5()` will default to the uniform 5th order coefficients ("uniform" setting) in all directions with a JS-WENO formulation
- `WENO5(grid = grid)` will default to uniform interpolation coefficient if the direction is uniform (`typeof(Δc) <: Number`) and will precompute the ENO coefficients for reconstruction in the stretched directions ("stretched" setting) (I have tested "on the fly" coefficient calculation for stretched directions and it is way too expensive! therefore, useless to keep as a possibility)
- `WENO5(grid = grid, stretched_smoothness=true)` will compute coefficients for the smoothness indicators `β₀, β₁` and `β₂` to account for stretched grid
- `WENO5(zweno = true)` will implement a Z-WENO formulation for the weno weight calculation
- No support is given for `WENO5S(grid = grid)` for curvilinear grids for the moment (defaults to uniform setting)

Comments:
Despite the fact that all methods have the same execution speed, `stretched_smoothness` requires more memory (and slightly more computation time) and is not much impactful. As such, most of the time it is better to use just the `WENO5(grid = grid)` keyword argument as it does not decrease accuracy but decreases memory utilization (and speed up slightly). (I haven't tried all types of grids, so maybe it is good to check before performing a large simulation on a weird stretched grid)
On the other hand, a Z-WENO formulation is always beneficial (also in case of a uniform mesh) with no major decrease in performance. The same can be said for the stretched `WENO5(grid=grid)` formulation in case of stretched grids

`validation/advection/validate_weno_scheme.jl` compares all these methods in terms of time and accuracy on a simple 1D and 2D tracer advection simulations. Below some animations showing the performance of the three methods for a "center coarsened" grid type (`grid_str2`)
`U` => `WENO5()` 
`S` => `WENO5(grid = grid)` 
`β` => `WENO5(grid = grid, stretched_smoothness=true)` 
`Z` => `WENO5(grid = grid, stretched_smoothness=true, zweno=true)` 

https://user-images.githubusercontent.com/7112768/142819747-4e4083cf-d725-4f1a-bac7-4fb10800ecc7.mp4

https://user-images.githubusercontent.com/7112768/142819755-fec13fb6-684f-4f1a-b7ec-78d9768c4ff7.mp4

Closes #1704

### CODE REVIEW

**Comment by tomchor** [comment]
Awesome PR!

This closes https://github.com/CliMA/Oceananigans.jl/issues/1704

I'll take the liberty to add this statement to the first post so that it actually closes it once this is merged.

**Comment by simone-silvestri** [comment]
By the way, i didn't use the scheme described in #1704 because it didn't give great results when implemented. I have implemented the ENO coefficients as derived before the assumption of uniform grid to the stencil interpolation (tomorrow I'll go back and check the report which describes them) . The smoothness functions are kept the same

**Review by navidcy — COMMENTED** [review]
I don't understand why don't we just replace the current WENO5 with the WENO5S and just call the new WENO5S simply WENO5?

Doesn't the WENO5S work for both regular and stretched grids?

**Comment by simone-silvestri** [comment]
Yeah sure, we can do it

**Comment by tomchor** [comment]
> By the way, i didn't use the scheme described in #1704 because it didn't give great results when implemented. I have implemented the ENO coefficients as derived before the assumption of uniform grid to the stencil interpolation (tomorrow I'll go back and check the report which describes them) . The smoothness functions are kept the same

Ah, interesting. If you can find the report or a source for that the we can cite/link in the docstring, I think it would be nice :)

**Comment by francispoulin** [comment]
This is a great PR, thanks @simone-silvestri for putting it together!

There is a script [here ](https://github.com/CliMA/Oceananigans.jl/blob/main/validation/advection/plot_rates_convergence_advection.jl)that computes the convergence rates fro the different advection schemes.  Given your results I'm sure that replacing `WENO5` by your new new version, would give the same results.  

But this script uses a uniform grid. Would it be of interest trying this with a streteched grid to see what order of convergence we get with the different schemes?

File: test/validate_weno_scheme.jl [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+using BenchmarkTools
+using Test
+using LinearAlgebra
+
+"""
+
+This simulation is a simple 1D advection of a gaussian function, to test the 
+validity of the stretched WENO scheme
+
+cannot make this work unless the pressure solver is set for a non-regular grid (which does not exist for the moment)
```
Comment (francispoulin) [inline test/validate_weno_scheme.jl:15]: What do you mean here exactly?  What will not work?

**Review by francispoulin — APPROVED** [review]
Great changes!

**Comment by simone-silvestri** [comment]
@tomchor this is the report from Shu deriving ENO coefficients for a finite volume scheme https://www3.nd.edu/~zxu2/acms60790S13/Shu-WENO-notes.pdf  (In particular equation 2.20 which becomes equation 2.21 for uniform grids)

@francispoulin Cool! I'll check the convergence then. What in the comments is that I use a Nonhydrostatic model which computes pressure and there is no infrastructure in Nonhydrostratic.jl for a pressure solver on a non-uniform horizontal grid (it will throw an error "no method Pressuresolver(grid::RectilinearGrid)". I maybe should change that test to an hydrostatic model...

**Comment by tomchor** [comment]
> By the way, it would be quite easy to modify the type to include the order of the WENO scheme which we could change at will. What do you think? Would it be usefull to do that?

If it's easy, then I definitely vote that we do that! A lower order, faster WENO scheme may be a good compromise between computational cost and accuracy. If that's true, we could use it as the default scheme for models.

**Comment by francispoulin** [comment]
First, I think it would be great to have multi-order `WENO` advection, if it's not too difficult.

Second, you can see an older version of the plot [here](https://github.com/CliMA/Oceananigans.jl/pull/1648).   Since then we've added first order upwinding.  I haven't tried running this for a while and hope it works but I guess I should try it and find out.

I should point out that this script uses `ShallowWaterModel`.  It is easy enough to change it for another of the two models, but I did see some odd behaviour.  For `ShallowWaterModel` we have that the error goes down to 1e-14, basically machine precision and the slopes are what we should expect.  When I tried this for `NonhydrostaticModel`, I found that the error didn't get any better than 1e-11 or so.  I looked into this but never figured out why.  The major difference between the two models is that `ShallowWater` does no thave a pressure solve, so I suspect that might be part of the problem.  In the end I decided that the latter was probably the best way to test pure advection.  But in retrospect, if we used `PrescribedVelocity`, then the pressure solve would not be a factor, and things should work out nicely.

**Comment by glwagner** [comment]
> Yeah sure, we can do it

Is there a performance difference?

The best solution might use dispatch to enable "non-stretched WENO" automatically when the grid has constant spacing.

**Comment by glwagner** [comment]
Can we call `src/Advection/weno_fifth_order_stretched.jl` just `src/Advection/weno_fifth_order.jl`?

File: src/Advection/weno_fifth_order_stretched.jl [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+    
+    if grid isa Nothing
+        coeff_xᶠᵃᵃ = nothing 
+        coeff_xᶜᵃᵃ = nothing  
+        coeff_yᵃᶠᵃ = nothing   
+        coeff_yᵃᶜᵃ = nothing
+        coeff_zᵃᵃᶠ = nothing
+        coeff_zᵃᵃᶜ = nothing
+    elseif !(typeof(grid) <: RectilinearGrid)
+        @warn "Stretched WENO is not supported with grids other than Rectilinear, defaulting to Uniform WENO"
```
Comment (glwagner) [inline src/Advection/weno_fifth_order_stretched.jl:44]: Is there a technical reason why this won't work on curvilinear grids, or is it merely something that needs to be implemented still?

File: src/Advection/weno_fifth_order_stretched.jl [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+function WENO5(FT = Float64; grid = nothing) 
+    
+    if grid isa Nothing
+        coeff_xᶠᵃᵃ = nothing 
+        coeff_xᶜᵃᵃ = nothing  
+        coeff_yᵃᶠᵃ = nothing   
+        coeff_yᵃᶜᵃ = nothing
+        coeff_zᵃᵃᶠ = nothing
+        coeff_zᵃᵃᶜ = nothing
+    elseif !(typeof(grid) <: RectilinearGrid)
```
Comment (glwagner) [inline src/Advection/weno_fifth_order_stretched.jl:43]:
```suggestion
    elseif !(grid isa RectilinearGrid)
```

Comment (glwagner) [reply src/Advection/weno_fifth_order_stretched.jl:43]: This same pattern applies below.

File: src/Advection/weno_fifth_order_stretched.jl (hunk 2 above)
Comment (simone-silvestri) [reply src/Advection/weno_fifth_order_stretched.jl:44]: I still have to research how to change coefficients with a curvilinearly stretched grid (I guess the vertical direction would be fine)

File: src/Grids/rectilinear_grid.jl [hunk 4]
```diff
@@ -355,19 +355,19 @@ function with_halo(new_halo, old_grid::RectilinearGrid)
     if old_grid.Δxᶠᵃᵃ isa Number
         x = x_domain(old_grid)
     else
-        x = old_grid.xᶠᵃᵃ
+        x = OffsetArray(Array(parent(old_grid.xᶠᵃᵃ)), old_grid.xᶠᵃᵃ.offsets[1])
```
Comment (glwagner) [inline src/Grids/rectilinear_grid.jl:358]:
```suggestion
        x = xnodes(Face, grid)
```

could also add convenience that a _range_ specification like `x = 0:0.1:10` in the constructor implies constant spacing. Then you can do away with the `if`-statement entirely here.

Comment (glwagner) [reply src/Grids/rectilinear_grid.jl:358]: I'm also not sure why this is needed --- can you explain? Why is it not enough to supply the _interior_ nodes to the constructor? I think the grid should be continued outside the domain according to some rule (we shouldn't allow users to set grid spacings outside the domain?)

File: src/Advection/weno_fifth_order_stretched.jl [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+        coeff_zᵃᵃᶜ = nothing    
+    else
+        FT          = eltype(grid)
+        arch        = grid.architecture
+        helper_grid = with_halo((4, 4, 4), grid)
+        if typeof(grid) <: XRegRectilinearGrid 
+            coeff_xᶠᵃᵃ = nothing 
+            coeff_xᶜᵃᵃ = nothing 
+        else
+            coeff_xᶠᵃᵃ = calc_interpolating_coefficients(FT, helper_grid.xᶠᵃᵃ, arch, grid.Nx)
```
Comment (glwagner) [inline src/Advection/weno_fifth_order_stretched.jl:59]: Might be _slightly_ simpler to define 

```julia
calc_interpolating_coefficients(FT, ::XRegRectilinearGrid, arch, grid.Nx) = nothing
```

then you don't need the if statement. Same for below.

Comment (simone-silvestri) [reply src/Advection/weno_fifth_order_stretched.jl:59]: Hmmm then you would need to define a separate `calc_interpolating_coefficients` for x-, y- and z-directions

Comment (glwagner) [reply src/Advection/weno_fifth_order_stretched.jl:59]:
```julia
calc_interpolating_coefficients(FT, ::XRegRectilinearGrid, arch, grid.Nx) = nothing

calc_interpolating_coefficients(FT, ::XRegRectilinearGrid, arch, grid.Nx) = nothing

Comment (glwagner) [reply src/Advection/weno_fifth_order_stretched.jl:59]: how about

```julia
calc_interpolating_coefficients(FT, ::OffsetArray{<:Any, <:AbstractRange}, arch, grid.Nx) = nothing
```

File: src/Grids/rectilinear_grid.jl (hunk 4 above)
Comment (simone-silvestri) [reply src/Grids/rectilinear_grid.jl:358]: It is just  way to ensure that the faces provided to the constructor are CPU arrays

File: src/Advection/weno_fifth_order_stretched.jl (hunk 5 above)
Comment (simone-silvestri) [reply src/Advection/weno_fifth_order_stretched.jl:59]: Sure, that works!

File: src/Grids/rectilinear_grid.jl (hunk 4 above)
Comment (glwagner) [reply src/Grids/rectilinear_grid.jl:358]: Ok, so perhaps `Array(xnodes(Face, grid))`?

File: src/Advection/weno_fifth_order_stretched.jl (hunk 2 above)
Comment (glwagner) [reply src/Advection/weno_fifth_order_stretched.jl:44]: I missed that you did not implement the algorithm in [A simple algorithm to improve the performance of the WENO scheme on non-uniform grids](https://link.springer.com/article/10.1007/s10409-017-0715-2) (which reports results for curvilinear grids). Perhaps we will have to find another way then.

**Comment by glwagner** [comment]
> By the way, i didn't use the scheme described in #1704 because it didn't give great results when implemented. I have implemented the ENO coefficients as derived before the assumption of uniform grid to the stencil interpolation (tomorrow I'll go back and check the report which describes them) . The smoothness functions are kept the same

That's interesting. Does that mean you were not able to reproduce the results reported in "[A simple algorithm to improve the performance of the WENO scheme on non-uniform grids](https://link.springer.com/article/10.1007/s10409-017-0715-2)"?

> > By the way, it would be quite easy to modify the type to include the order of the WENO scheme which we could change at will. What do you think? Would it be usefull to do that?
> 
> If it's easy, then I definitely vote that we do that! A lower order, faster WENO scheme may be a good compromise between computational cost and accuracy. If that's true, we could use it as the default scheme for models.

I think a higher order WENO scheme is more important than lower-order scheme. However, a low-order WENO method is proposed [here](https://link.springer.com/article/10.1007/s10915-020-01164-6).

#995 attempts to implement nth order WENO but I think failed to achieve good CPU performance / compilation on the GPU. But @simone-silvestri perhaps you're able to achieve this?

**Comment by navidcy** [comment]
I could have a look on this PR. @simone-silvestri could you update the first comment or is it reflecting what's in here?

**Comment by simone-silvestri** [comment]
@navidcy I have changed a couple of settings. It's almost ready, I'll update in an hour or so

File: validation/advection/plot_rates_convergence_advection.jl [hunk 6]
```diff
@@ -60,19 +60,48 @@ schemes = (
...
+        if i < N/4
+         return 1
+        elseif i > N/4*3
+         return 1
+        elseif i<N/2
+         return 1.2 * (i - N/4) + 1
+        else
+         return  1.2 * (3*N/4 - i) + 1
+        end
+     end
```
Comment (navidcy) [inline validation/advection/plot_rates_convergence_advection.jl:83]:
```suggestion
    function Δx_str2(i, N)
        if i < N/4
            return 1
        elseif i > 3N/4
            return 1
        elseif i < N/2
            return 1.2 * (i - N/4) + 1
        else
            return 1.2 * (3N/4 - i) + 1
        end
     end   
```

File: src/Oceananigans.jl [hunk 7]
```diff
@@ -20,7 +20,7 @@ export
     xnodes, ynodes, znodes, nodes,
 
     # Advection schemes
-    CenteredSecondOrder, CenteredFourthOrder, UpwindBiasedFirstOrder, UpwindBiasedThirdOrder, UpwindBiasedFifthOrder, WENO5,
+    CenteredSecondOrder, CenteredFourthOrder, UpwindBiasedFirstOrder, UpwindBiasedThirdOrder, UpwindBiasedFifthOrder, WENO5,
```
Comment (navidcy) [inline src/Oceananigans.jl:23]:
```suggestion
    CenteredSecondOrder, CenteredFourthOrder, UpwindBiasedFirstOrder, UpwindBiasedThirdOrder, UpwindBiasedFifthOrder, WENO5,

```

**Comment by simone-silvestri** [comment]
Finished. It should work for all the settings on CPU and GPU.

File: src/Advection/weno_fifth_order.jl [hunk 8]
```diff
@@ -177,17 +177,27 @@ Adapt.adapt_structure(to, scheme::WENO5{FT, XT, YT, ZT, XS, YS, ZS, W}) where {F
         stencil = retrieve_left_smooth(scheme, r, dir, i, location)
         wᵢᵢ = stencil[1]   
         wᵢⱼ = stencil[2]
+        # horrible but have to do this for GPU execution (broadcast doesn't work apparently)
```
Comment (glwagner) [inline src/Advection/weno_fifth_order.jl:180]: You can't broadcast _inside_ a kernel. Broadcasting can _launch_ a kernel. Everything inside a kernel has to compile to CUDA C. I didn't know that the compile could unroll loops like what you've written though, that's awesome!

**Review by navidcy — COMMENTED** [review]
few beautifications

File: src/Advection/weno_fifth_order.jl (hunk 8 above)
Comment (simone-silvestri) [reply src/Advection/weno_fifth_order.jl:180]: Ah cool! Launching a kernel with a broadcast seems like a nice feature! (Tool me a while to figure it out though:) ) Then you mean that while running on the GPU, threads at lines 152 -154 are launching their own child dynamic kernel with size 3? Or is the compiler handling it in some weird way?

Regarding the for loops from what I know there is nothing preventing them in CUDA C

**Comment by navidcy** [comment]
@simone-silvestri could you point me to a test/validation that confirms 1) that the changes of the PR didn't have an effect on regular grids. Also, a benchmark to see how this PR might have speedup/slowed down things? There were some benchmarks in the first comment when the PR was opened but many changes were pursued after. Should we repeat those benchmarks or something along those lines?

**Comment by simone-silvestri** [comment]
@navidcy Good call! 
You can find here below the output of benchmarks/benchmark_advection_schemes.jl ran on main and this PR and adapted to test the new methods.
For validation, you can run validation/advection/validate_weno_schemes.jl to validate the methods on a 1D and 2D advection problem for different grids
(uniform and stretched)

execution of benchmark_advection_schemes.jl on main
```
┌───────────────┬────────────────────────┬───────────┬───────────┬───────────┬───────────┬──────────┬────────┬─────────┐
│ Architectures │                Schemes │       min │    median │      mean │       max │   memory │ allocs │ samples │
├───────────────┼────────────────────────┼───────────┼───────────┼───────────┼───────────┼──────────┼────────┼─────────┤
│           CPU │    CenteredFourthOrder │   1.779 s │   1.789 s │   1.794 s │   1.814 s │ 1.05 MiB │   1685 │       3 │
│           CPU │    CenteredSecondOrder │   1.018 s │   1.066 s │   1.061 s │   1.090 s │ 1.05 MiB │   1685 │       5 │
│           CPU │ UpwindBiasedFifthOrder │   2.467 s │   2.505 s │   2.505 s │   2.544 s │ 1.05 MiB │   1685 │       2 │
│           CPU │ UpwindBiasedThirdOrder │   1.808 s │   1.877 s │   1.862 s │   1.903 s │ 1.05 MiB │   1685 │       3 │
│           CPU │                  WENO5 │   6.775 s │   6.775 s │   6.775 s │   6.775 s │ 1.05 MiB │   1685 │       1 │
│           GPU │    CenteredFourthOrder │ 20.879 ms │ 20.970 ms │ 21.078 ms │ 21.874 ms │ 2.17 MiB │  20610 │      10 │
│           GPU │    CenteredSecondOrder │ 11.402 ms │ 15.533 ms │ 15.138 ms │ 15.680 ms │ 2.05 MiB │  13147 │      10 │
│           GPU │ UpwindBiasedFifthOrder │ 23.356 ms │ 23.498 ms │ 24.055 ms │ 29.246 ms │ 2.21 MiB │  23281 │      10 │
│           GPU │ UpwindBiasedThirdOrder │ 18.863 ms │ 19.027 ms │ 19.298 ms │ 21.745 ms │ 2.12 MiB │  17945 │      10 │
│           GPU │                  WENO5 │ 23.234 ms │ 28.467 ms │ 27.982 ms │ 28.684 ms │ 2.30 MiB │  29259 │      10 │
└───────────────┴────────────────────────┴───────────┴───────────┴───────────┴───────────┴──────────┴────────┴─────────┘
```
This PR same file, (WENO5 on uniform grids)
```
┌───────────────┬────────────────────────┬───────────┬───────────┬───────────┬───────────┬──────────┬────────┬─────────┐
│ Architectures │                Schemes │       min │    median │      mean │       max │   memory │ allocs │ samples │
├───────────────┼────────────────────────┼───────────┼───────────┼───────────┼───────────┼──────────┼────────┼─────────┤
│           CPU │    CenteredFourthOrder │   1.758 s │   1.763 s │   1.762 s │   1.766 s │ 1.05 MiB │   1685 │       3 │
│           CPU │    CenteredSecondOrder │   1.018 s │   1.027 s │   1.034 s │   1.053 s │ 1.05 MiB │   1685 │       5 │
│           CPU │ UpwindBiasedFifthOrder │   2.335 s │   2.339 s │   2.343 s │   2.353 s │ 1.05 MiB │   1685 │       3 │
│           CPU │ UpwindBiasedThirdOrder │   1.787 s │   1.892 s │   1.867 s │   1.922 s │ 1.05 MiB │   1685 │       3 │
│           CPU │                  WENO5 │   6.935 s │   6.935 s │   6.935 s │   6.935 s │ 1.05 MiB │   1685 │       1 │
│           GPU │    CenteredFourthOrder │ 20.856 ms │ 21.016 ms │ 21.146 ms │ 22.351 ms │ 2.14 MiB │  18750 │      10 │
│           GPU │    CenteredSecondOrder │ 11.201 ms │ 15.574 ms │ 15.175 ms │ 15.932 ms │ 2.05 MiB │  13243 │      10 │
│           GPU │ UpwindBiasedFifthOrder │ 23.296 ms │ 23.514 ms │ 25.463 ms │ 43.065 ms │ 2.18 MiB │  21868 │      10 │
│           GPU │ UpwindBiasedThirdOrder │ 18.859 ms │ 19.041 ms │ 19.338 ms │ 22.157 ms │ 2.10 MiB │  16584 │      10 │
│           GPU │                  WENO5 │ 22.645 ms │ 28.377 ms │ 27.818 ms │ 28.571 ms │ 2.26 MiB │  27075 │      10 │
└───────────────┴────────────────────────┴───────────┴───────────┴───────────┴───────────┴──────────┴────────┴─────────┘
```
to test the new WENO methods on a stretched grid modify the code in benchmark_advection_schemes.jl as seen below

```
function benchmark_advection_scheme(Arch, Scheme)
    grid = RectilinearGrid(size=(192, 192, 192), x = (0, 1), y = (0, 1), z = collect(0:192), architecture = Arch())

    if Scheme == :Uniform
        scheme = WENO5()
    elseif Scheme == :Stretched
        scheme = WENO5(grid = grid)
    elseif Scheme == :StretchedSmoothness
        scheme = WENO5(grid = grid, stretched_smoothness = true)
    elseif Scheme == :ZFormulation
        scheme = WENO5(zweno = true)
    end
    model = NonhydrostaticModel(architecture=Arch(), grid=grid, advection=scheme)

    time_step!(model, 1) # warmup

    trial = @benchmark begin
        @sync_gpu time_step!($model, 1)
    end samples=10

    return trial
end

Schemes =  (:Uniform, :Stretched, :StretchedSmoothness, :ZFormulation)
```
This PR, with a "stretched" vertical direction 
```
┌───────────────┬─────────────────────┬────────────┬────────────┬────────────┬────────────┬────────────┬────────┬─────────┐
│ Architectures │             Schemes │        min │     median │       mean │        max │     memory │ allocs │ samples │
├───────────────┼─────────────────────┼────────────┼────────────┼────────────┼────────────┼────────────┼────────┼─────────┤
│           CPU │           Stretched │    7.122 s │    7.122 s │    7.122 s │    7.122 s │ 949.91 KiB │   1716 │       1 │
│           CPU │ StretchedSmoothness │   11.042 s │   11.042 s │   11.042 s │   11.042 s │ 956.47 KiB │   1716 │       1 │
│           CPU │             Uniform │    6.373 s │    6.373 s │    6.373 s │    6.373 s │ 945.38 KiB │   1706 │       1 │
│           CPU │        ZFormulation │    7.725 s │    7.725 s │    7.725 s │    7.725 s │ 945.38 KiB │   1706 │       1 │
│           GPU │           Stretched │  25.646 ms │  25.920 ms │  26.501 ms │  32.101 ms │   2.15 MiB │  30442 │      10 │
│           GPU │ StretchedSmoothness │ 109.080 ms │ 109.493 ms │ 109.676 ms │ 111.118 ms │   3.78 MiB │ 134069 │      10 │
│           GPU │             Uniform │  25.133 ms │  25.394 ms │  25.468 ms │  26.710 ms │   2.06 MiB │  27799 │      10 │
│           GPU │        ZFormulation │  30.554 ms │  32.780 ms │  32.569 ms │  32.922 ms │   2.20 MiB │  37382 │      10 │
└───────────────┴─────────────────────┴────────────┴────────────┴────────────┴────────────┴────────────┴────────┴─────────┘
```
Apparently on a much larger grid than what I tested, StretchedSmoothness is much less efficient, another reason to try to avoid it if not strictly necessary!

**Comment by navidcy** [comment]
OK. All these is great!

So all this info you just shared is it anywhere else written? E.g. "StretchedSmoothness is much less efficient... another reason to avoid it". Btw, what exactly is "StretchedSmoothness"?

We should have a docs section on advection schemes (#1318) and perhaps this PR is not the right place to deal with that.

**Review by navidcy — APPROVED** [review]
I haven't gone through the code changes line by line, but the benchmarks + validations convince me that nothing was broken in the process. 

So, given the above disclaimer, I'm approving!

Ideally we'd like somewhere some info or a little cheatsheet regarding whether we should use which WENO5 flavour.

**Comment by navidcy** [comment]
I pushed some minor edits to the validation script.

**Comment by navidcy** [comment]
I also couldn't play the movies in the first comment. I suspect because they were created with `gif(anim, "movie.mp4", ...)`. I re-run the validation script and replaced them.

**Comment by simone-silvestri** [comment]
@navidcy thanks for the review! 
Indeed maybe we should document it. 

Just to give a quick explanation of the different flavours

WENO involves the reconstruction of the value of the advected quantity on the face on different "biased stencils"



where  is the reconstruction polynomial constructed from the biased stencil  and the  is the order of the polynomial. As we want a 5th order scheme, there will be three polynomials () of order 2 (). Now the coefficients of these polynomials () are based on the Lagrange interpolation of the integral of  and, as such, are coordinate-dependent

The first thing to do, when wanting to implement the WENO scheme on a stretched grid, is then to make sure that the reconstruction is correctly done, and, as such, `WENO5(grid = grid)` takes care that ()  are correctly calculated based on the underlying grid.

Now, WENO schemes differ from ENO schemes since they do not just choose the "smoothest stencil" among the (in this case) three stencils, but weight them as such



Those weights have to satisfy  and are mostly a matter of choice (if the reconstruction is performed correctly for all stencils!)

The general way to do this is to make them functions of a "local smoothness indicator"   which is calculated as such
 

These  simplify quite nicely for uniform grids. the `streched_smoothness=true` option calculates the coefficients of these smoothness indicators for a stretched grid, by assuming that ()

Then the weights are calculated as such
 

where  depends on the scheme used and differs between JS-WENO and Z-WENO. In the former

In the latter 


 are optimized constants equal in both the schemes.

Now, I believe that `stretched_smoothness` does not have a such big impact because if you assure that the reconstruction is correct, then all stencil correctly interpolate to the face and weights will still be between 0 and 1. This might be much more impactful for grids that are much more non-uniform than what we usually use.

**Comment by tomchor** [comment]
Great explanation, @simone-silvestri! Just to make sure I understand some practical aspects:

- WENO with `streched_smoothness=false` is like the WENO we had before?
- WENO with `streched_smoothness=true` increases the accuracy of WENO (by using custom, pre-calculated coefficients for the grid), but there's a non-negligible speed penalty?

**Comment by simone-silvestri** [comment]
@tomchor 

In terms of "smoothness indicators" yes, not in terms of polynomial reconstruction. The WENO we had before is `WENO5()`. This will be formally wrong on a stretched grid because the reconstruction at the face is not done accounting for variable spacing. 

`WENO5(grid = grid, stretched_smoothness=false) ` accounts for variable spacing in the reconstruction polynomial (i.e., the  coefficients are calculated based on grid spacing). This makes the scheme formally correct as, independently on how you weight the polynomials, the different stencils will have mathematically correct reconstructions on the face. This should _always_ be done in case of a stretched grid (and it does not decrease execution speed).

Now, the `stretched_smoothness=true` calculates the  accounting for a stretched mesh. This means that the three reconstruction polynomials () are weighted accounting for their actual smoothness and not a "uniform smoothness" that `WENO5(grid = grid, stretched_smoothness=false) ` calculates. In my tests this did not give a noticeable benefit in terms of accuracy (apparently  for a stretched mesh are very similar to   for a uniform mesh). In addition, it gives a non-negligible decrease in execution speed (especially on the GPU). For this reason, I think it is nice to have as an option, but to be used with parsimony :) and only when necessary.

File: validation/advection/plot_rates_convergence_advection.jl [hunk 9]
```diff
@@ -60,19 +60,48 @@ schemes = (
  UpwindBiasedThirdOrder(), 
  CenteredFourthOrder(), 
  UpwindBiasedFifthOrder(), 
+ WENO5(),
  WENO5()
```
Comment (tomchor) [inline validation/advection/plot_rates_convergence_advection.jl:64]: Is this from then there was `WENO5` and `WENO5S`?

**Review by tomchor — APPROVED** [review]

File: validation/advection/plot_rates_convergence_advection.jl [hunk 10]
```diff
@@ -60,19 +60,48 @@ schemes = (
  UpwindBiasedThirdOrder(), 
  CenteredFourthOrder(), 
  UpwindBiasedFifthOrder(), 
+ WENO5(),
```
Comment (navidcy) [inline validation/advection/plot_rates_convergence_advection.jl:63]:
```suggestion
```

**Comment by navidcy** [comment]
> @tomchor
> 
> In terms of "smoothness indicators" yes, not in terms of polynomial reconstruction. The WENO we had before is `WENO5()`. This will be formally wrong on a stretched grid because the reconstruction at the face is not done accounting for variable spacing.
> 
> `WENO5(grid = grid, stretched_smoothness=false) ` accounts for variable spacing in the reconstruction polynomial (i.e., the  coefficients are calculated based on grid spacing). This makes the scheme formally correct as, independently on how you weight the polynomials, the different stencils will have mathematically correct reconstructions at the face. This should _always_ be done in case of a stretched grid (and it does not decrease execution speed).
> 
> Now, the `stretched_smoothness=true` calculates the  accounting for a stretched mesh. This means that the three reconstruction polynomials () are weighted accounting for their actual smoothness and not a "uniform smoothness" that `WENO5(grid = grid, stretched_smoothness=false) ` calculates. In my tests this did not give a noticeable benefit in terms of accuracy (apparently  for a stretched mesh are very similar to  for a uniform mesh). In addition, it gives a non-negligible decrease in execution speed (especially on the GPU). For this reason, I think it is nice to have as an option, but to be used with parsimony :) and only when necessary.

OK. This is great. Do you mind if I add it in a docstring somewhere? It's a pity to be left just in a comment in this PR, lost in the aether of closed PRs...

**Comment by simone-silvestri** [comment]
Cool, thanks for documenting

**Comment by navidcy** [comment]
I noticed "On the other hand, a Z-WENO formulation is *always* beneficial (also in case of a uniform mesh) with no major
decrease in performance."

Why don't we then change `zweno = true`?

File: src/Advection/weno_fifth_order.jl [hunk 11]
```diff
@@ -1,134 +1,367 @@
...
+(The above claims were made after some preliminary tests. Thus, we still users to perform some
+benchmarks/checks before performing, e.g., a large simulation on a "weirdly" stretched grid.)
+
+On the other hand, a Z-WENO formulation is *always* beneficial (also in case of a uniform mesh) with no major
+decrease in performance. The same can be said for the stretched `WENO5(grid = grid)` formulation in case of
+stretched grids.
+
+References
+==========
+
```
Comment (navidcy) [inline src/Advection/weno_fifth_order.jl:165]: @simone-silvestri, I didn't include all the info you provided in the PR comments in the docstring. I felt that what I included was helpful enough. But perhaps here we add 2-3 relevant references?

Comment (simone-silvestri) [reply src/Advection/weno_fifth_order.jl:131]: @navidcy, thanks for the work. I'll add a couple of references

**Comment by simone-silvestri** [comment]
@navidcy I have made a couple of small changes. Anyways, I am a little hesitant to make Z-WENO the default as it does, yes, have slight benefits, but we tested only on small simplified problems. I would have a bit more testing before making it the default.

File: src/Advection/weno_fifth_order.jl [hunk 12]
```diff
@@ -1,12 +1,210 @@
...
+On the other hand, a Z-WENO formulation is *most of the times* beneficial (also in case of a uniform mesh) with roughly the 
+same performances (just a slight slow/down). The same can be said for the stretched `WENO5(grid = grid)` formulation in case of
+stretched grids.
+
+References
+==========
+
+Shu, Essentially Non-Oscillatory and Weighted Essentially Non-Oscillatory Schemes for Hyperbolic Conservation Laws, 1997, NASA/CR-97-206253, ICASE Report No. 97-65
+Castro et al, High order weighted essentially non-oscillatory WENO-Z schemesfor hyperbolic conservation laws, 2011, Journal of Computational Physics 230(5):1766-1792
+
```
Comment (navidcy) [inline src/Advection/weno_fifth_order.jl:134]:
```suggestion
- Shu, Essentially Non-Oscillatory and Weighted Essentially Non-Oscillatory Schemes for Hyperbolic Conservation Laws, 1997, NASA/CR-97-206253, ICASE Report No. 97-65
- Castro et al, High order weighted essentially non-oscillatory WENO-Z schemes for hyperbolic conservation laws, 2011, Journal of Computational Physics 230(5):1766-1792
```

**Comment by navidcy** [comment]
> @navidcy I have made a couple of small changes. Anyways, I am a little hesitant to make Z-WENO the default as it does, yes, have slight benefits, but we tested only on small simplified problems. I would have a bit more testing before making it the default.

Sounds good! I was just wondering.

I made a formatting suggestion. I think you should merge as soon as tests pass!

**Comment by simone-silvestri** [comment]
Cool! I'll merge!

**Pull request merged**